File: Tyche.Android/app/src/main/java/com/pipel/tyche/pool/useCase/FindPoolsUseCase.kt
package com.pipel.tyche.pool.useCase

import com.pipel.core.CursorPage
import com.pipel.tyche.pool.data.PoolRepository
import com.pipel.tyche.pool.domain.Pool
import com.pipel.tyche.pool.mapper.PoolMapper
import javax.inject.Inject

interface FindPoolsUseCase {

    suspend fun execute(
        poolLayoutId: String,
        nextToken: String?,
        filter: String?
    ): CursorPage<Pool>

}

class DefaultFindPoolsUseCase @Inject constructor(private val poolRepository: PoolRepository) :
    FindPoolsUseCase {

    override suspend fun execute(
        poolLayoutId: String,
        nextToken: String?,
        filter: String?
    ): CursorPage<Pool> {
        return poolRepository.getPools(poolLayoutId, nextToken, filter)
            .map(PoolMapper::mapFromDataToDomain)
    }

}

File: Tyche.Backend/Pipel.Tyche.Pool.Api/Pipel.Tyche.Pool.Api/FindPoolsUseCase.fs
module Pipel.Tyche.Pool.Api.FindPools

open Pipel.Core
open Pipel.Data
open Pipel.Tyche.Pool.Domain
open Pipel.Tyche.Pool.Domain.UseCases
open Pipel.Type

let execute
    (poolLayoutId: string)
    (filter: string)
    (next: string)
    (findPoolsUseCase: IFindPoolsUseCase)
    (mapFromDomainToViewFunc: MapFunc<Pool, PoolResponse>)
    =
    async {
        let! page =
            findPoolsUseCase.AsyncExecute(
                { PoolLayoutPK.PoolLayoutId = Ulid.From poolLayoutId },
                filter |> String.tryFrom,
                next |> String.tryFrom
            )

        return page |> CursorPage.map mapFromDomainToViewFunc
    }
